repo/blob/logging/logging_storage.go

```diff
@@ -3,19 +3,41 @@ package logging
 
 import (
 	"context"
+	"sync/atomic"
 
 	"github.com/kopia/kopia/internal/timetrack"
 	"github.com/kopia/kopia/repo/blob"
 	"github.com/kopia/kopia/repo/logging"
 )
 
 type loggingStorage struct {
+	concurrency    int32
+	maxConcurrency int32
+
 	base   blob.Storage
 	prefix string
 	logger logging.Logger
 }
 
+func (s *loggingStorage) beginConcurrency() {
+	v := atomic.AddInt32(&s.concurrency, 1)
+
+	if mv := atomic.LoadInt32(&s.maxConcurrency); v > mv {
+		if atomic.CompareAndSwapInt32(&s.maxConcurrency, mv, v) {
+			s.logger.Debugw(s.prefix+"concurrency level reached",
+				"maxConcurrency", mv)
```

Comment:
minor:

Is there a _small_ race here?

Suppose:
* the state is `concurrency == maxConcurrency`, for example `7`;
* there are two routines `g1` and `g2` concurrently running `beginConcurrency`;
* `g1` gets `v == 8` and `g2` gets `v == 9`;
* both `g1` and `g2` get `mv == 7` and both attempt to update `maxConcurrency`;
* `g1`'s update succeeds and now `maxConcurrency == 8`, but `concurrency > maxConcurrency`

This may self correct if there is a higher concurrency value down the road.